File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.cpp
class Solution {
public:
    int minimumPushes(string word) {
        int n = word.size();
        if(n<8) return n;
        int arr[26]={0}, c=0;
        for(int i=0; i<n; i++){
            arr[word[i]-'a']++;
        }
        sort(arr, arr+26, greater<int>());
        for(int i=0; i<26; i++){
            c+=arr[i]*((i/8)+1);
        }
        return c;
    }
};

File: 3016-minimum-number-of-pushes-to-type-word-ii/3016-minimum-number-of-pushes-to-type-word-ii.py
from collections import Counter
class Solution:
    def minimumPushes(self, word: str) -> int:
        frequencies = Counter(word)
        sorted_freqs = sorted(frequencies.values(), reverse=True)
        
        pushes = 0
        for index, freq in enumerate(sorted_freqs):
            pushes_per_letter = (index // 8) + 1
            pushes += freq * pushes_per_letter
            
        return pushes
